### PR TITLE
Improve Process tests

### LINKED ISSUE #1874: System.Diagnostics.ProcessTests.ProcessTest.Process_AsyncHalfCharacterAtATime test sometimes fails
System.Diagnostics.ProcessTests.ProcessTest.Process_AsyncHalfCharacterAtATime [FAIL]
       Assert.True() Failure
       Expected: True
       Actual:   False
       Stack Trace:
          E:\oss\corefx\src\System.Diagnostics.Process\tests\System.Diagnostics.Process.Tests\Process_StreamTests.cs(131,0):
          at System.Diagnostics.ProcessTests.ProcessTest.Process_AsyncHalfCharacterAtATime()

### PR DESCRIPTION
**Pull request by Maxwe11**
also fixes #1874

### CODE REVIEW

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/Process_StreamTests.cs [hunk 1]
```diff
@@ -42,17 +43,34 @@ public void Process_SyncErrorStream()
...
+            }
 
-            if (p.WaitForExit(WaitInMS))
-                p.WaitForExit(); // This ensures async event handlers are finished processing.
+            {
+                // Now add the CancelAsyncAPI as well.
+                StringBuilder sb = new StringBuilder();
+                Process p = CreateProcessError();
+                p.StartInfo.RedirectStandardError = true;
+                p.ErrorDataReceived += (s, e) => { sb.Append(e.Data); ((Process)s).CancelErrorRead(); };
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/Process_StreamTests.cs]: What's being tested here?

And if the only difference between the two versions is whether CancelErrorRead is called, did you consider just putting the whole test in a loop for the two cases, executing the CancelErrorRead call on just one iteration?

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/Process_StreamTests.cs [hunk 2]
```diff
@@ -116,19 +134,89 @@ public void Process_SyncStreams()
...
+
+                Assert.Throws<InvalidOperationException>(() => p.StandardOutput);
+                Assert.Throws<InvalidOperationException>(() => p.StandardError);
+
+                if (p.WaitForExit(WaitInMS))
+                    p.WaitForExit(); // This ensures async event handlers are finished processing.
+            }
+
+            {
+                Process p = CreateProcessStream();
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/Process_StreamTests.cs]: I must just be missing something... is this block not identical to the previous one?

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/Process_StreamTests.cs (hunk 1 above)
Comment (Maxwe11) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/Process_StreamTests.cs]: Picked this test case from similar for `CancelOutputRead()` below. Will update it as well.

**Comment by stephentoub** [comment]
Thanks! LGTM

**Comment by stephentoub** [comment]
@pallavit, look ok to you?

**Comment by pallavit** [comment]
Thanks. LGTM :-)

**Pull request merged**